Sources/SwiftProtobuf/BytecodeReader.swift

```diff
@@ -0,0 +1,150 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+
+/// Reads values encoded in a SwiftProtobuf bytecode stream.
+package struct BytecodeReader<Instruction: RawRepresentable> where Instruction.RawValue == UInt64 {
+    /// The remaining slice of the program that has not yet been read.
+    private var remainingProgram: UnsafeBufferPointer<UInt8>.SubSequence
+
+    /// Indicates whether or not there is still data that hasn't yet been read in the bytecode
+    /// stream.
+    package var hasData: Bool {
+        !remainingProgram.isEmpty
+    }
+
+    /// Creates a new bytecode reader that reads the given bytecode stream.
+    package init(remainingProgram: UnsafeBufferPointer<UInt8>.SubSequence) {
+        self.remainingProgram = remainingProgram
+
+        // We reserve the first integer of the program text as a "format specifier". This
+        // future-proofs us if we ever want to change the way programs themselves are encoded
+        // (for example, compressing them).
+        let programFormat = nextUInt64()
+        if programFormat > latestBytecodeProgramFormat {
+            fatalError("Unexpected bytecode program format \(programFormat)")
+        }
+    }
+
+    /// Reads and returns the next instruction from the bytecode stream.
+    ///
+    /// - Precondition: The reader must not be at the end of the bytecode stream, and the next
+    ///   opcode must not be zero.
+    ///
+    /// - Returns: The instruction that was read from the bytecode stream.
+    package mutating func nextInstruction() -> Instruction {
+        precondition(hasData, "Unexpected end of bytecode stream")
+
+        let opcode = nextUInt64()
+        precondition(opcode != 0, "Opcode 0 is reserved; do not use it in your own instructions")
+        guard let instruction = Instruction(rawValue: opcode) else {
+            fatalError("Unexpected opcode \(opcode) for instruction set \(Instruction.self)")
+        }
+        return instruction
+    }
+
+    /// Reads and returns the next signed 32-bit integer from the bytecode stream.
+    ///
+    /// This is provided as its own primitive operation because 32-bit values are extremely common
+    /// as field numbers (0 to 2^29-1) and enum cases (-2^31 to 2^31-1). In particular for enum
+    /// cases, using this function specifically for those cases avoids making mistakes involving
+    /// sign- vs. zero-extension between differently-sized integers.
+    ///
+    /// - Precondition: The reader must not be at the end of the bytecode stream.
+    ///
+    /// - Returns: The signed 32-bit integer that was read from the bytecode stream.
+    package mutating func nextInt32() -> Int32 {
+        // `Int32`s are stored by converting them bit-wise to a `UInt32` and then zero-extended to
```

Comment:
Should this assert if the value was more than 32 bits?  I'm thinking of some safety net to ensure even negative numbers don't get expanded to64 bits.